OSDOCS#6057: Authenticating cert-manager Operator on the GCP (without workload identity) (#60433)
openshift/openshift-docs: modules/cert-manager-configure-cloud-credentials-gcp-non-sts.adoc

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-non-sts-gcp.adoc

:_content-type: PROCEDURE
[id="cert-manager-prepare-cloud-credentials-gcp-non-sts"]
= Configuring cloud credentials for the {cert-manager-operator} on GCP

> You can configure cloud credentials for the {cert-manager-operator} on GCP.

Reviewer: ```suggestion
In order to configure the cloud credentials for the {cert-manager-operator} on a GCP cluster you need to create a `CredentialsRequest` object and allow OpenShift Cloud Credential Operator to generate the cloud credentials secret.
```

.Prerequisites

> * You must extract and prepare the `ccoctl` binary.

Reviewer: ```suggestion
```
We can remove ccoctl dependency, as we do not need it for non-workload identity GCP.

* The {cert-manager-operator} 1.11.1 or later is installed.

.Procedure

. Create a directory to store the `CredentialsRequest` custom resource (CR) of the {cert-manager-operator} by running the following command:
+
[source,terminal]
----
$ mkdir credentials-request
----

. Change the directory by running the following command:
+
[source,terminal]
----
$ cd credentials-request

> ----

Reviewer: Ditto: https://github.com/openshift/openshift-docs/pull/59907/files#r1209215583
we can skip steps 1, 2 and directly move to what's in step 3

  generation: 1
  labels:
    app: cert-manager
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: cert-manager
    app.kubernetes.io/name: cert-manager
  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: GCPProviderSpec
    predefinedRoles:
    - roles/dns.admin
  secretRef:
    name: gcp-credentials
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

. Update the subscription object for {cert-manager-operator} by running the following command:
+
[source,terminal]
----
$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"gcp-credentials"}]}}}'
----

. Apply the secrets generated in the manifests directory of your cluster by running the following command:
+
[source,terminal]
----
$ ls manifests/*-credentials.yaml | xargs -I{} oc apply -f {}
----

.Verification

> . Verify that the secret is updated by running the following command:
> +
> [source,terminal]
> ----
> $ oc get deployment cert-manager -n cert-manager -o yaml
> ----
> +
> .Example output
> [source,yaml]
> ----
> - name: cloud-credentials
>         secret:
>           defaultMode: 420
>           items:
>           - key: service_account.json
>             path: application_default_credentials.json
>           secretName: gcp-credentials
> ----

Reviewer: ```suggestion
```
You can remove this part and keep what's in the latter.

. Get the name of a cert-manager controller pod by running the following command:
+
[source,terminal]
----
$ oc get pods -l app.kubernetes.io/name=cert-manager -n cert-manager
----
+
.Example output
[source,terminal]
----
NAME                                       READY   STATUS    RESTARTS   AGE
cert-manager-bd7fbb9fc-wvbbt               1/1     Running   0          15m39s
----

. Verify that GCP credentials are updated by running the following command:
+
[source,terminal]
----
$ oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o=jsonpath='{.spec.volumes}'
----
+
.Example output
[source,terminal]
----

> 

Reviewer: ```suggestion
- name: cloud-credentials
  secret:
    defaultMode: 420
    items:
    - key: service_account.json
      path: application_default_credentials.json
    secretName: gcp-credentials
```
Also, you can add the output of spec.volume here instead.

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-non-sts-gcp.adoc

:_content-type: PROCEDURE
[id="cert-manager-prepare-cloud-credentials-gcp-non-sts"]
= Configuring cloud credentials for the {cert-manager-operator} on GCP

To configure the cloud credentials for the {cert-manager-operator} on a GCP cluster you must create a `CredentialsRequest` object and allow the Cloud Credential Operator to generate the cloud credentials secret.

.Prerequisites

> * The {cert-manager-operator} 1.11.1 or later is installed.

Author: How about: `You have configured the CCO mode to use administrator-level cloud credentials in your cluster. `

> * The {cert-manager-operator} 1.11.1 or later is installed.

Reviewer: I was leaning more towards something like:

```
Your cluster is running Cloud Credential Operator in either Mint or Passthrough mode.
```

> * The {cert-manager-operator} 1.11.1 or later is installed.

Author: Updated the prereq to `You have configured the Cloud Credential Operator to operate in _mint_ or _passthrough_ mode.`

.Procedure

. Create a `CredentialsRequest` resource YAML file, for example, `sample-credential-request.yaml`, as follows:
+
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  generation: 1
  labels:
    app: cert-manager
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: cert-manager

>     app.kubernetes.io/name: cert-manager

Reviewer 2: Not needed.

Reviewer: @xenolinux If you end up removing the metadata.labels in the CredentialsRequest, please do the same in the other 3 PRs as well.

  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: GCPProviderSpec
    predefinedRoles:
    - roles/dns.admin
  secretRef:
    name: gcp-credentials
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

. Update the subscription object for {cert-manager-operator} by running the following command:
+
[source,terminal]
----

> $ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"gcp-credentials"}]}}}'

Reviewer 2: ```suggestion
$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type=merge -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"gcp-credentials"}]}}}'
```

----

.Verification

> . Get the name of a cert-manager controller pod by running the following command:

Reviewer 2: ```suggestion
. Get the name of the redeployed cert-manager controller pod by running the following command:
```

+
[source,terminal]
----
$ oc get pods -l app.kubernetes.io/name=cert-manager -n cert-manager
----
+
.Example output
[source,terminal]
----
NAME                                       READY   STATUS    RESTARTS   AGE
cert-manager-bd7fbb9fc-wvbbt               1/1     Running   0          15m39s
----

. Verify that GCP credentials are updated by running the following command:
+
[source,terminal]
----

> $ oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o=jsonpath='{.spec.volumes}'

Reviewer 2: Same with another PR, this will be incorrect. -o=jsonpath='{.spec.volumes}' will print JSON format, but your Example output is YAML format.

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-non-sts-gcp.adoc

:_content-type: PROCEDURE

> [id="cert-manager-prepare-cloud-credentials-gcp-non-sts"]

Reviewer 3: ```suggestion
[id="cert-manager-prepare-cloud-credentials-gcp-non-sts_{context}"]
```

= Configuring cloud credentials for the {cert-manager-operator} on GCP

To configure the cloud credentials for the {cert-manager-operator} on a GCP cluster you must create a `CredentialsRequest` object and allow the Cloud Credential Operator to generate the cloud credentials secret.

.Prerequisites

* The {cert-manager-operator} 1.11.1 or later is installed.
* You have configured the Cloud Credential Operator to operate in _mint_ or _passthrough_ mode.

.Procedure

> . Create a `CredentialsRequest` resource YAML file, for example, `sample-credential-request.yaml`, as follows:

Reviewer 3: ```suggestion
. Create a `CredentialsRequest` resource YAML file, such as, `sample-credential-request.yaml` by applying the following yaml:
```

  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: GCPProviderSpec
    predefinedRoles:
    - roles/dns.admin
  secretRef:
    name: gcp-credentials
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

. Update the subscription object for {cert-manager-operator} by running the following command:
+
[source,terminal]
----
$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type=merge -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"gcp-credentials"}]}}}'
----

.Verification

. Get the name of the redeployed cert-manager controller pod by running the following command:
+
[source,terminal]
----
$ oc get pods -l app.kubernetes.io/name=cert-manager -n cert-manager
----
+
.Example output
[source,terminal]
----
NAME                                       READY   STATUS    RESTARTS   AGE
cert-manager-bd7fbb9fc-wvbbt               1/1     Running   0          15m39s
----

> . Verify that the cert-manager controller pod is updated with GCP workload identity credential volumes mounted under the path specified in `mountPath` by running the following command:

Reviewer 3: ```suggestion
. Verify that the cert-manager controller pod is updated with GCP workload identity credential volumes  that are mounted under the path specified in `mountPath` by running the following command:
```

+
[source,terminal]
----
$ oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o yaml
----
+
.Example output
[source,terminal]
----
spec:
  containers:
  - args:
    ...
    volumeMounts:
    ...
    - mountPath: /.config/gcloud
      name: cloud-credentials
    ....
  volumes:
  ...
  - name: cloud-credentials
    secret:
      ...
      items:
      - key: service_account.json
        path: application_default_credentials.json

>       secretName: cloud-credentials

Author: @xingxingxia Addressed

To configure the cloud credentials for the {cert-manager-operator} on a GCP cluster you must create a `CredentialsRequest` object and allow the Cloud Credential Operator to generate the cloud credentials secret.

.Prerequisites

* The {cert-manager-operator} 1.11.1 or later is installed.

.Procedure

. Create a `CredentialsRequest` resource YAML file, for example, `sample-credential-request.yaml`, as follows:
+
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  generation: 1
  labels:
    app: cert-manager
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: cert-manager
    app.kubernetes.io/name: cert-manager
  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: GCPProviderSpec
    predefinedRoles:
    - roles/dns.admin
  secretRef:
    name: gcp-credentials
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

> . Apply the secrets generated in the manifests directory of your cluster by running the following command:

Reviewer: Missed it in earlier reviews, this step is also not required and we should remove it.

+
[source,terminal]
----
$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"gcp-credentials"}]}}}'
----

. Apply the secrets generated in the manifests directory of your cluster by running the following command:
+
[source,terminal]
----
$ ls manifests/*-credentials.yaml | xargs -I{} oc apply -f {}

> ----

Author: Addressed
